stl/inc/regex

```diff
@@ -3398,19 +3398,48 @@ bool _Matcher<_BidIt, _Elem, _RxTraits, _It>::_Do_class(_Node_base* _Nx) { // ap
 }
 
 template <class _BidIt, class _Elem, class _RxTraits, class _It>
-bool _Matcher<_BidIt, _Elem, _RxTraits, _It>::_Better_match() { // check for better match under UNIX rules
+bool _Matcher<_BidIt, _Elem, _RxTraits, _It>::_Better_match() {
+    // check for better match under the rules of the current regex implementation
     for (unsigned int _Ix = 0; _Ix < _Get_ncap(); ++_Ix) { // check each capture group
-        if (_Res._Grp_valid[_Ix] && _Tgt_state._Grp_valid[_Ix]) {
-            if (_Res._Grps[_Ix]._Begin != _Tgt_state._Grps[_Ix]._Begin) {
-                return _STD distance(_Begin, _Res._Grps[_Ix]._Begin)
-                     < _STD distance(_Begin, _Tgt_state._Grps[_Ix]._Begin);
+        const auto& _Res_grp = _Res._Grps[_Ix];
+        const auto& _Tgt_grp = _Tgt_state._Grps[_Ix];
+        if (_Res_grp._Begin == _Res_grp._End) {
+            if (_Tgt_grp._Begin == _Tgt_grp._End) {
+                // Both groups are empty or invalid - matches are equivalent.
+                continue;
             }
+            // Target group is valid and non-empty, and current group is either empty or invalid.
+            return true;
+        }
+        if (_Tgt_grp._Begin == _Tgt_grp._End) {
+            // Target group is invalid or empty, and current group is valid and non-empty,
+            // so our current group is the better match.
+            return false;
+        }
 
-            if (_Res._Grps[_Ix]._End != _Tgt_state._Grps[_Ix]._End) {
-                return _STD distance(_Begin, _Res._Grps[_Ix]._End) < _STD distance(_Begin, _Tgt_state._Grps[_Ix]._End);
+        if (_Res_grp._Begin != _Tgt_grp._Begin) {
+            ptrdiff_t _Res_dist_from_begin = _STD distance(_Begin, _Res_grp._Begin);
+            ptrdiff_t _Tgt_dist_from_begin = _STD distance(_Begin, _Tgt_grp._Begin);
+            if (_Sflags & regex_constants::ECMAScript) {
+                // If one of the matches starts before the other, in ECMAScript, it is considered better.
+                // This is also known as "leftmost", in languages written from left to right.
+                return _Tgt_dist_from_begin < _Res_dist_from_begin;
+            } else {
+                // Otherwise, Unix uses "rightmost" rules.
+                return _Res_dist_from_begin < _Tgt_dist_from_begin;
             }
         }
+
+        if (_Res_grp._End != _Tgt_grp._End) {
+            // If one of the matches is longer than the other, and they start at the same point,
+            // it is considered better.
+            ptrdiff_t _Res_size = _STD distance(_Res_grp._Begin, _Res_grp._End);
+            ptrdiff_t _Tgt_size = _STD distance(_Tgt_grp._Begin, _Tgt_grp._End);
+            return _Tgt_size > _Res_size;
+        }

```

Comment:
I believe this logic is incorrect (although I don't immediately know what all of the problems are).

First, although it is not an authoritative reference, https://en.cppreference.com/w/cpp/regex/syntax_option_type describes POSIX as "leftmost longest". I'm surprised to see "rightmost" mentioned in this implementation. (From what I've read, ECMAScript is not "leftmost longest".) Boost explains the Perl/ECMAScript behavior at https://www.boost.org/doc/libs/1_80_0/libs/regex/doc/html/boost_regex/syntax/perl_syntax.html#boost_regex.syntax.perl_syntax.what_gets_matched and the POSIX behavior at https://www.boost.org/doc/libs/1_80_0/libs/regex/doc/html/boost_regex/syntax/leftmost_longest_rule.html .

Second, the cppreference article depicts the output:
```
Searching for .*(a|xayy) in zzxayyzz:
 ECMA (depth first search) match: zzxa
 POSIX (leftmost longest)  match: zzxayy
```

However, both `main` and this PR print:
```
Searching for .*(a|xayy) in zzxayyzz:
 ECMA (depth first search) match: zzxayy
 POSIX (leftmost longest)  match: zzxayy
```

In this example with this PR, the final `if`-statement is saying "they start at the same point, so prefer the longer one".

(As an aside, this PR's logic is structured and commented in a clear, comprehensible way, which is great! My only concern is matching what the grammars specify.)

Reply:
hmm, yeah, I apparently misunderstood something as I was cleaning up the logic; I think the existing logic was worse, but yeah; not sure what the best way forward is.